### ISSUE TITLE
投擲回転制御

### ISSUE THREAD

**Issue by TaisukeOtsuki** [CITRobocon/RC2018#3]

**Comment by TaisukeOtsuki** [comment]
#include<mbed.h>
#include<QEI.h>
#include<iostream>
#include "mo.h"
QEI encode(D10,D14,D15,1000,QEI::X2_ENCODING);

namespace robocon{
    void Motors::read_angle(void){
       float count=encode.getPulses()*(-1);
       float rev =encode.getRevolutions();
       float angle=encoder_2000_360(count);
       if(rev==1){
           encode.reset();
           angle=0;
       }
       ang=angle;
       printf("angle=%f\r\n",angle);
    }        
    Motors::Motors(PinName motor_pin, int Period, int ZeroPoint, int Max, int Min)
           : mt(motor_pin), max_pulse(Max), min_pulse(Min), zero_pulse(ZeroPoint)  {          
        mt.period_ms(Period);
        mt.pulsewidth_us(ZeroPoint);
        read_angle();
        flag=0;
        ang=0;
        before=0;
        
            
    }   
    float Motors::encoder_2000_360(float kakudo/*360*/){
        return 360.0f*kakudo/2000.0f;
    }
    float Motors::adjust_move(float move){          
        return  (500/100*move)+zero_pulse;
    }
    void Motors::motor_move(int move){
        int spam=move-before;
        int mv;
        mv=before;
        if(spam>=40){
            while(1){
                mv++;
                mt.pulsewidth_us(adjust_move(mv)); 
                wait(0.05);
                if(mv==move){break;}
            }
        }
        if(spam<=-40){
            while(1){
            mv--;
            mt.pulsewidth_us(adjust_move(mv)); 
            wait(0.05);
            if(mv==move){break;}   
            }           
        }           
        mt.pulsewidth_us(adjust_move(mv));    
        wait(0);
        before=move;
        v=before;
    }
    void Motors::P_motion(float Q){
        int t;
        float q;
        float K=70;
        while(1){
            read_angle();
            
        }    
      motor_move(0);         
    }    
    void Motors::control_P(float _Q,float p) {
        
        read_angle();
        motor_move(p);
        wait(2.0);
        printf("ang=%d,_Q=%f\r\n",ang,_Q);
        while(1){
            read_angle();
            if(ang==_Q){
                P_motion(_Q);
                break;
            }
        
        }
    }      
    
}

**Comment by sanoshohei** [comment]
'''cpp
#include "mbed.h"
#include "QEI.h"

BusIn in(D10, D14, D15);
QEI enc(D10,D14,D15,1000, QEI::X2_ENCODING);
PwmOut motor(PA_15);
Serial pc(USBTX,USBRX);

#define pw_zero 1500
#define pw_max 1700
#define pw_min 1300
#define bottom 20

Ticker control;

double delta_t = 0.020;
double preang, ang, t_ang;
double duty_motor;
int pw_motor = pw_zero;
int cnt_enc;
double k1 = 1.5, k2 = 400.0;

void control_motor(){
    cnt_enc = enc.getPulses()*(-1);
    ang = 360*(double)cnt_enc/2000;
    
    if(enc.getRevolutions() != 0){
      enc.reset();
      printf ("ang = %lf\n\r", ang);
    }
 
    if (t_ang-ang > 180)
      duty_motor = ((t_ang-ang-360)*k1 - (ang-preang)*delta_t*k2)/180;
    else if (t_ang-ang < -180)
      duty_motor = ((t_ang-ang+360)*k1 - (ang-preang)*delta_t*k2)/180;
    else
      duty_motor = ((t_ang-ang)*k1 - (ang-preang)*delta_t*k2)/180;
    
    if (duty_motor < 0.0)
      pw_motor = duty_motor*500 + pw_zero - bottom;
    else
      pw_motor = duty_motor*500 + pw_zero + bottom;
    
    if (pw_motor > pw_max)
      pw_motor = pw_max;
    else if (pw_motor < pw_min)
      pw_motor = pw_min;
    
    motor.pulsewidth_us(pw_motor);
    
    if (preang == ang)
      printf ("%lf\n\r", ang);
    
    preang = ang;
}
    
int main(){
    in.mode(PullUp);
    motor.period_ms(10);
    motor.pulsewidth_us(1540);
    enc.reset();
    while(enc.getRevolutions() != 1);
    enc.reset();
    motor.pulsewidth_us(1500);
    printf ("\n\rcompleted!\n\r");
    wait(3.0);
    preang = enc.getPulses()*(-1)*360/2000;
    printf ("start!\n\r");
    control.attach(&control_motor, delta_t);
    t_ang = 90;
    wait(3.0);
    t_ang = 270;
    wait(5.0);
    t_ang = 90;
}

**Comment by sanoshohei** [comment]
#include "mbed.h"
#include "QEI.h"

BusIn in(D10, D14, D15);
QEI enc(D10,D14,D15,1000, QEI::X2_ENCODING);
PwmOut motor(PA_15);
Serial pc(USBTX,USBRX);

#define ppr 2000

#define er_ang 0.050

#define delta_t 0.020

#define pw_zero 1500
#define pw_maxran 500
#define pw_ran 200
#define bottom 5

Ticker control;

double preang, ang, t_ang;
double duty_motor;
int pw_motor = pw_zero;
int cnt_enc;


void control_motor(){
    ang = enc.getPulses()*(-1);
    if (ang < 0)
      ang += 2000;
    else if (ang > 2000)
      ang -= 2000;
    
    if(enc.getRevolutions())
      enc.reset();
 
    if (t_ang-ang > ppr/2.0)
      duty_motor = (t_ang-ang-ppr)/(ppr/2.0);
    else if (t_ang-ang < -ppr/2.0)
      duty_motor = (t_ang-ang+ppr)/(ppr/2.0);
    else
      duty_motor = (t_ang-ang)/(ppr/2.0);
    
    if (duty_motor < 0.0)
      pw_motor = duty_motor*pw_maxran + pw_zero - bottom;
    else
      pw_motor = duty_motor*pw_maxran + pw_zero + bottom;
    
    if (pw_motor > pw_zero+pw_maxran)
      pw_motor = pw_zero+pw_maxran;
    else if (pw_motor < pw_zero-pw_maxran)
      pw_motor = pw_zero-pw_maxran;
    
    motor.pulsewidth_us(pw_motor);
    
    preang = ang;
}


void setup_enc (void){
    motor.pulsewidth_us(pw_zero+pw_ran/4);
    enc.reset();
    while(!enc.getRevolutions());
    enc.reset();
    motor.pulsewidth_us(pw_zero);
    preang = enc.getPulses()*(-1);
    printf("setuped\n\r");
    
    return;
}


void set_ang (int t_deg){
    t_ang = (double)t_deg / 360 * 2000;
    control.attach(&control_motor, delta_t);
    while(ang > t_ang*(1.0+er_ang) || ang < t_ang*(1.0-er_ang) || ang != preang);
    wait(5.0);
    control.detach();
    motor.pulsewidth_us(pw_zero);
    printf("setted degree\n\r");
    
    return;
}

    
int main(){
    in.mode(PullUp);
    motor.period_ms(10);
    setup_enc();
    wait(1.0);
    set_ang(90);
    set_ang(0);
    while(1);
}
